File: sol/40/40.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        n = len(candidates)
        nextIdx = [-1] * n
        i = 0
        j = 0
        while i < n:
            while j < n and candidates[j] == candidates[i]:
                j += 1
            for k in range(i, j):
                nextIdx[k] = j
            i = j
        res = []
        def backtrack(remain, idx, comb):
            if remain == 0:
                nonlocal res
                res.append(comb.copy())
                return
            if idx == n or candidates[idx] > remain:
                return
            comb.append(candidates[idx])
            backtrack(remain - candidates[idx], idx + 1, comb)
            comb.pop()
            backtrack(remain, nextIdx[idx], comb)
        backtrack(target, 0, [])
        return res

File: sol/40/40.cpp
struct Solution {
    void backtrack(int remain, int idx, vector<int> &comb, vector<vector<int>> &res, vector<int> &nextIdx, vector<int> &candidates){
        if (remain == 0){
            res.push_back(comb);
            return;
        }
        if (idx == candidates.size() || candidates[idx] > remain){
            return;
        }
        comb.push_back(candidates[idx]);
        this->backtrack(remain - candidates[idx], idx + 1, comb, res, nextIdx, candidates);
        comb.pop_back();
        this->backtrack(remain, nextIdx[idx], comb, res, nextIdx, candidates);
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        ranges::sort(candidates);
        int n = candidates.size();
        vector<int> nextIdx(n);
        for (int i = 0, j = 0; i < n; i = j){
            while (j < n && candidates[j] == candidates[i]){
                ++j;
            }
            for (int k = i; k < j; ++k){
                nextIdx[k] = j;
            }
        }
        vector<vector<int>> res;
        vector<int> comb;
        this->backtrack(target, 0, comb, res, nextIdx, candidates);
        return res;
    }
};
